examples/bindings/quickstart_sparse_string/Program.cs

```diff
@@ -0,0 +1,321 @@
+            var end_handle = GCHandle.Alloc(row_end, GCHandleType.Pinned);
+            Methods.tiledb_query_add_range_var_by_name(ctx, read_query, rows_name,
+            start_handle.AddrOfPinnedObject().ToPointer(), (ulong)row_start.Length,
+            end_handle.AddrOfPinnedObject().ToPointer(), (ulong)row_end.Length
+            );
+            check_err(ctx);
+
+            int cols_start = 2, cols_end = 4;
+            Methods.tiledb_query_add_range(ctx, read_query, 1, &cols_start, &cols_end, null);
+            check_err(ctx);
+
+            Methods.tiledb_query_set_layout(ctx, read_query, tiledb_layout_t.TILEDB_ROW_MAJOR);
+            check_err(ctx);
+
+            var read_data_handle = GCHandle.Alloc(read_data, GCHandleType.Pinned);
+            var read_data_size_handle = GCHandle.Alloc(read_data_size, GCHandleType.Pinned);
+            Methods.tiledb_query_set_data_buffer(ctx, read_query, attr_name,
+                read_data_handle.AddrOfPinnedObject().ToPointer(), (ulong*)read_data_size_handle.AddrOfPinnedObject()
+            );
+            read_data_handle.Free();
+            read_data_size_handle.Free();
+            check_err(ctx);
+
+            read_data_handle = GCHandle.Alloc(read_rows, GCHandleType.Pinned);
+            read_data_size_handle = GCHandle.Alloc(read_rows_size, GCHandleType.Pinned);
+            Methods.tiledb_query_set_data_buffer(ctx, read_query, rows_name,
+                read_data_handle.AddrOfPinnedObject().ToPointer(), (ulong*)read_data_size_handle.AddrOfPinnedObject()
+            );
+            read_data_handle.Free();
+            read_data_size_handle.Free();
+            check_err(ctx);
+
+            read_data_handle = GCHandle.Alloc(read_rows_offsets, GCHandleType.Pinned);
+            read_data_size_handle = GCHandle.Alloc(read_rows_offsets_size, GCHandleType.Pinned);
+            Methods.tiledb_query_set_offsets_buffer(ctx, read_query, rows_name,
+                (ulong*)read_data_handle.AddrOfPinnedObject(), (ulong*)read_data_size_handle.AddrOfPinnedObject()
+            );
+            read_data_handle.Free();
+            read_data_size_handle.Free();
+            check_err(ctx);
+
+            read_data_handle = GCHandle.Alloc(read_cols, GCHandleType.Pinned);
+            read_data_size_handle = GCHandle.Alloc(read_cols_size, GCHandleType.Pinned);
+            Methods.tiledb_query_set_data_buffer(ctx, read_query, cols_name,
+                read_data_handle.AddrOfPinnedObject().ToPointer(), (ulong*)read_data_size_handle.AddrOfPinnedObject()
+            );
+            read_data_handle.Free();
+            read_data_size_handle.Free();
+            check_err(ctx);
+
+            Methods.tiledb_query_submit(ctx, read_query);
+            check_err(ctx);
+
+            for (int i = 0; i < read_data.Length; i++)
+            {
+                // No more data to print if offset is 0 after first iteration
+                if (i > 0 && read_rows_offsets[i] == 0) break;

```

Comment:
We should not rely on checking against 0 here, because this read might be out of bounds. Instead, when the `tiledb_query_submit` call returns above, the memory location pointed to by `read_data_size` (note: needs to be separate for each `set_..._buffer` call) will contain the total bytes read for that particular buffer. For offsets, we can divide by `sizeof(uint64_t)` to find the number of cells read. See the `buffer_size` doc comment here: https://tiledb-inc-tiledb.readthedocs-hosted.com/en/stable/c-api.html#c.tiledb_query_set_buffer

Finally, note that the size of the last cell must be calculated by subtracting the last offset from the total array size.